### PR TITLE
Added KITTI dataset

### PR DESCRIPTION
**Pull request by prabhat00155**
Discussion: https://github.com/pytorch/vision/issues/3562

### CODE REVIEW

**Comment by prabhat00155** [comment]
I see tests failing due to pandas import error. Should I add it to config.yml?

File: test/test_datasets.py [hunk 1]
```diff
@@ -155,6 +160,19 @@ def test_places365_repr_smoke(self):
             dataset = torchvision.datasets.Places365(root, download=True)
             self.assertIsInstance(repr(dataset), str)
 
+    def test_kitti(self):
```
Comment (NicolasHug) [inline test/test_datasets.py:163]: We're trying to move away from these tests and instead rely on `ImageDatasetTestCase` as the other classes below in this file. Would you mind porting it there?

File: torchvision/datasets/kitti.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
+import os
+from collections import namedtuple
+from typing import Any, Callable, NamedTuple, Optional, Tuple
+from urllib.error import URLError
+
+import pandas as pd
```
Comment (NicolasHug) [inline torchvision/datasets/kitti.py:6]: instead of importing paandas here, let's import it in `__init__` as done for the `Celeba` dataset.

File: torchvision/datasets/kitti.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+from PIL import Image
+
+from .utils import download_and_extract_archive
+from .vision import VisionDataset
+
+
+class Kitti(VisionDataset):
+    """`KITTI <http://www.cvlibs.net/datasets/kitti>`_ Dataset.
```
Comment (NicolasHug) [inline torchvision/datasets/kitti.py:14]: We should also mention the pandas dependency here in the docstring

**Review by NicolasHug — COMMENTED** [review]
Thanks @prabhat00155 , made a quick review.

Regarding the dependency on pandas: once the test is ported to use `ImageDatasetTestCase`, you can set `    REQUIRED_PACKAGES = ("pandas",)`. Feel free to take a look at the `CelebATestCase` class which also relies on `pd.read_csv`.

File: torchvision/datasets/kitti.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                score: FloatTensor[N]
+        """
+        image = Image.open(self.images[index])
+        target = None if self.split == "test" else self._parse_target(index)
+        if self.transforms:
+            image, target = self.transforms(image, target)
+        if self.transform:
+            image = self.transform(image)
+        if self.target_transform:
+            target = self.target_transform(target)
```
Comment (fmassa) [inline torchvision/datasets/kitti.py:130]: Instead of having both `transforms` as well as `transform` and `target_transform`, I propose to instead only have a single `transforms` call.

In the current implementation, we would be calling the transforms twice, as `VisionDataset` provides an automatic replacement for `transform` and `transforms` in https://github.com/pytorch/vision/blob/f9af70a98244b30f645fb8a033b6e4ce59ca020f/torchvision/datasets/vision.py#L27-L29

This is also what other datasets similar to this one (like COCO and VOC) do, see https://github.com/pytorch/vision/blob/f9af70a98244b30f645fb8a033b6e4ce59ca020f/torchvision/datasets/coco.py#L48-L49 for example

File: torchvision/datasets/kitti.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                "rotation_y",
+            ],
+        )
+        self.images = []
+        self.targets = []
+        self.root = root
+        self.split = split
+        self.transform = transform
+        self.target_transform = target_transform
+        self.transforms = transforms
```
Comment (fmassa) [inline torchvision/datasets/kitti.py:86]: This is already taken care of by `VisionDataset` constructor, so no need to do it again (and it will in fact not work as `VisionDataset` was originally expected to do

**Review by fmassa — COMMENTED** [review]
The PR looks great, thanks a lot @prabhat00155 !

I've made a few more comments, let me know what you think

File: torchvision/datasets/kitti.py (hunk 3 above)
Comment (prabhat00155) [reply torchvision/datasets/kitti.py:14]: Removed pandas dependency.

File: torchvision/datasets/kitti.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            content = csv.reader(inp, delimiter=" ")
+            for line in content:
+                target["type"].append(line[0])
+                target["truncated"].append(line[1])
+                target["occluded"].append(line[2])
+                target["alpha"].append(line[3])
+                target["bbox"].append(line[4:8])
+                target["dimensions"].append(line[8:11])
+                target["location"].append(line[11:14])
+                target["rotation_y"].append(line[14])
```
Comment (fmassa) [inline torchvision/datasets/kitti.py:125]: Quick question: does this for loop represent different objects?
If yes, I wonder which representation might make more sense to expose to the user by default (`List[Dict]` or `Dict[List]`). Do you have any thoughts on this?

Also, just wondering if `csv` automatically casts the data in `line` to be `int` / `float` / etc, or if the data is always represented as strings?

Comment (prabhat00155) [reply torchvision/datasets/kitti.py:125]: Yeah, this loop corresponds to different objects present in the image. That's a fair point, a list of dictionaries make more sense here.
`csv` leaves everything as strings. I think we should convert it into the right format. I'll update the code. Thanks!

**Comment by fmassa** [comment]
Windows test failures seem related, can you have a look?

File: torchvision/datasets/kitti.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                    "location": [float(x) for x in line[11:14]],
+                    "rotation_y": float(line[14]),
+                })
+        return target
+
+    def __len__(self) -> int:
+        return len(self.images)
+
+    @property
+    def raw_folder(self) -> str:
```
Comment (NicolasHug) [inline torchvision/datasets/kitti.py:132]: Unless this is expected to be public, I'd suggest to make it private: `_raw_folder`. By making it private we're protecting ourselves against the impossiblity of making BC breaking changes, as it happened in https://github.com/pytorch/vision/pull/3611#issuecomment-810352148

File: torchvision/datasets/kitti.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        location = "testing" if self.split == "test" else "training"
+        folders = [self.image_dir_name]
+        if self.split != "test":
+            folders.append(self.labels_dir_name)
+        return all(
+            os.path.isdir(os.path.join(self.raw_folder, location, fname))
+            for fname in folders
+        )
+
+    def download(self) -> None:
```
Comment (NicolasHug) [inline torchvision/datasets/kitti.py:146]: Same here, we can probably make this private (unless this is supposed to be common to all datasets?)

File: torchvision/datasets/__init__.py [hunk 9]
```diff
@@ -24,6 +24,7 @@
 from .hmdb51 import HMDB51
 from .ucf101 import UCF101
 from .places365 import Places365
+from .kitti import Kitti
```
Comment (NicolasHug) [inline torchvision/datasets/__init__.py:27]: I'd even suggest to name new files with a leading underscore, so that all objects defined in this file are private by default. But I'd confirm with @fmassa first before making this change.

```suggestion
from ._kitti import Kitti
```

**Review by NicolasHug — COMMENTED** [review]
Thanks @prabhat00155 ,

The tests look great!

I think you'll also need to add the dataset in `docs/source/datasets.rst` so that the docs are rendered properly. You can check the docs locally or in this PR: click on the `build_docs` CI run, and then "Artifacts" tab

Pinging @pmeier for another look ;)

**Review by fmassa — APPROVED** [review]
Looks great to me, thanks a lot @prabhat00155 !

File: torchvision/datasets/kitti.py (hunk 8 above)
Comment (fmassa) [reply torchvision/datasets/kitti.py:146]: I think `download` is currently a common method across all datasets, so I'd be fine with it

File: torchvision/datasets/__init__.py (hunk 9 above)
Comment (fmassa) [reply torchvision/datasets/__init__.py:27]: This is a good point. I think for the sake of consistency we should just keep `kitti` as is for now, but we can discuss this in an issue.

File: torchvision/datasets/kitti.py (hunk 7 above)
Comment (prabhat00155) [reply torchvision/datasets/kitti.py:132]: We may want to update `mnint.py` as well(I can create a separate PR for that).

File: test/datasets_utils.py [hunk 10]
```diff
@@ -44,7 +44,7 @@ class UsageError(Exception):
 
 
 class LazyImporter:
-    r"""Lazy importer for additional dependicies.
+    r"""Lazy importer for additional dependencies.
```
Comment (pmeier) [inline test/datasets_utils.py:47]: Good catch! Could you revert this here and send a follow-up PR in order to keep the scope small?

File: test/test_datasets.py [hunk 11]
```diff
@@ -1,18 +1,18 @@
...
 from torchvision.datasets import utils
-from common_utils import get_tmp_dir
-from fakedata_generation import svhn_root, places365_root, widerface_root, stl10_root
+from fakedata_generation import (
+    places365_root,
+    stl10_root,
+    svhn_root,
+)
 import xml.etree.ElementTree as ET
-from urllib.request import Request, urlopen
```
Comment (pmeier) [inline test/test_datasets.py:15]: Same as above. I know the imports are messy right now, but fixing this does not belong in this PR. I suggest we wait with this until #3531 is closed.

File: torchvision/datasets/kitti.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+            and its target as entry and returns a transformed version.
+        download (bool, optional): If true, downloads the dataset from the internet and
+            puts it in root directory. If dataset is already downloaded, it is not
+            downloaded again.
+
+    """
+
+    mirrors = [
+        "https://s3.eu-central-1.amazonaws.com/avg-kitti/",
+    ]
```
Comment (pmeier) [inline torchvision/datasets/kitti.py:45]: @fmassa Are we planning on using mirrors everywhere? If yes, I think we should implement this functionality in `download_url`. Otherwise I don't think we should start out with mirrors if the official server is stable.

File: torchvision/datasets/kitti.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        self.split = split
+
+        if download:
+            self.download()
+        if not self._check_exists():
+            raise RuntimeError(
+                "Dataset not found. You may use download=True to download it."
+            )
+
+        location = "testing" if self.split == "test" else "training"
```
Comment (pmeier) [inline torchvision/datasets/kitti.py:80]: This is also used in `_check_exist`. Maybe make it an instance attribute?
```suggestion
        self._location = "testing" if self.split == "test" else "training"
```

File: torchvision/datasets/kitti.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        "data_object_image_2.zip",
+        "data_object_label_2.zip",
+    ]
+    image_dir_name = "image_2"
+    labels_dir_name = "label_2"
+
+    def __init__(
+        self,
+        root: str,
+        split: str = None,
```
Comment (pmeier) [inline torchvision/datasets/kitti.py:56]: Not sure if I missed anything, but why are we using `None` as default value here? Shouldn't it be `"train"`? Plus, if there are only two splits, `train: bool = True` might even be easier. If I'm wrong and we need `split = None` somewhere, we should at least fix the annotation:

```suggestion
        split: Optional[str] = None,
```

File: test/test_datasets.py [hunk 15]
```diff
@@ -1702,5 +1702,43 @@ def test_classes(self, config):
...
+        os.makedirs(kitti_dir)
+
+        split_to_idx = split_to_num_examples = {
+            None: 1,
+            "train": 1,
+            "test": 2,
+        }
+
+        # We need to create all folders regardless of the split in config
+        for split in ("train", "test"):
```
Comment (pmeier) [inline test/test_datasets.py:1721]: This kinda confirms my suspicion from above. While we have  the ability to pass `split=None` it is not even used in our tests. Thus, I suggest we remove that.

File: test/test_datasets.py [hunk 16]
```diff
@@ -1702,5 +1702,43 @@ def test_classes(self, config):
...
+            split_idx = split_to_idx[split]
+            num_examples = split_to_num_examples[split]
+
+            datasets_utils.create_image_folder(
+                root=kitti_dir,
+                name=os.path.join(f"{split}ing", "image_2"),
+                file_name_fn=lambda image_idx: f"000{split_idx + image_idx}.png",
```
Comment (pmeier) [inline test/test_datasets.py:1728]: It is not terribly important here, but I have a hard time imagining that the file are named after this scheme in the original dataset. Wouldn't this mean that the second train image (`split_idx==1`, `image_idx==2`) and the first test image (`split_idx==2`, `image_idx==1`) have the same file name?

**Review by pmeier — CHANGES_REQUESTED** [review]
@prabhat00155 Looks good overall! I have a few comments below, though.

File: test/test_datasets.py (hunk 16 above)
Comment (prabhat00155) [reply test/test_datasets.py:1728]: The filenames in the KITTI dataset range from `000000.png` to `007480.png` in the training set and from `000000.png` to `007517.png` in the test set. Here in the test file, for training there will be one image created with `split_idx = 1, image_idx=0` and for test there will be two images with `split_idx=2, image_idx={0, 1}`, so shouldn't have the same file name. Training file name will be `0001.png` and testing file name will be `0002.png` and `0003.png`. Also, even if they were same, it won't matter as they are under different directories `training/image_2` and `testing/image_2`. And yes, I left out a few zeros while naming the files(compared to the actual dataset), which shouldn't really be significant in terms of test results.

Comment (pmeier) [reply test/test_datasets.py:1728]: > Also, even if they were same, it won't matter as they are under different directories `training/image_2` and `testing/image_2`

IMO we should stick to their file naming scheme. True, it doesn't make a difference here, but I stumbled over this during review. Given that they live in separate directories, we can simply use the same file name for every split.



> And yes, I left out a few zeros while naming the files

You can have auto zero padding:

```suggestion
                file_name_fn=lambda image_idx: f"{image_idx:06d}.png",
```

With this the number is padded with zeros in case it has less than 6 digits.

File: torchvision/datasets/kitti.py (hunk 12 above)
Comment (prabhat00155) [reply torchvision/datasets/kitti.py:45]: Perhaps I shouldn't have used the name `mirrors` here. This is the link they email you if you try to download the dataset. I kept it this way in case the link changes in future or there are multiple sources.

Comment (pmeier) [reply torchvision/datasets/kitti.py:45]: It is not about the name per se. Supporting multiple sources for the dataset makes the download logic more complex. So we should either use a central solution, i.e. by passing `mirrors` to `download_url` or don't implement it at all. Since we have no indication that this server is unstable for some reason, I would simply store a single URL and only use this. This way you need no looping, no error check, and so on.

**Comment by prabhat00155** [comment]
> @prabhat00155 Looks good overall! I have a few comments below, though.

@pmeier Addressed your comments, could you take a look? Thanks!

File: torchvision/datasets/kitti.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+        if self._check_exists():
+            return
+
+        os.makedirs(self._raw_folder, exist_ok=True)
+
+        # download files
+        for fname in self.resources:
+            url = f"{self.data_url}{fname}"
+            try:
```
Comment (pmeier) [inline torchvision/datasets/kitti.py:159]: There is no point in having this `try` / `except` block as well as the `print()` here anymore. `download_url()` handles this internally.

https://github.com/pytorch/vision/blob/07fb8ba7fad7b5b458ff862919825df4e6f60b52/torchvision/datasets/utils.py#L129-L139

**Review by pmeier — APPROVED** [review]
One minor thing left. Otherwise LGTM. Thanks a lot @prabhat00155!

File: torchvision/datasets/kitti.py (hunk 12 above)
Comment (fmassa) [reply torchvision/datasets/kitti.py]: I don't think we should add a `mirrors` functionality to `download_url` -- it just makes the logic more complicated.

**Comment by fmassa** [comment]
Thanks a ton @prabhat00155 !
